Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+```swift
+/// The base LambdaHandler protocol
+public protocol StreamingLambdaHandler: ~Copyable {
+    /// The business logic of the Lambda function
+    /// - Parameters:
+    ///   - event: The invocation's input data
+    ///   - responseWriter: A ``LambdaResponseWriter`` to write the invocations response to.
+    ///                     If no response or error is written to the `responseWriter` it will
+    ///                     report an error to the invoker.
+    ///   - context: The LambdaContext containing the invocation's metadata
+    mutating func handle(_ event: ByteBuffer, responseWriter: consuming some LambdaResponseWriter, context: LambdaContext) async
+}
+```
+
+The protocol is defined in a way that supports a broad range of use-cases. The handle method is marked as `mutating` to allow handlers to be implemented with a `struct`. Users can opt to even mark the handler as non-copyable.
+
+An implementation that sends the number 1 to 10 every 500ms could look like this:
+
+```swift
+struct SendNumbersWithPause: StreamingLambdaHandler {
+    func handle(
+        _ event: ByteBuffer,
+        responseWriter: consuming some LambdaResponseWriter,
+        context: LambdaContext
+    ) async {
+        for i in 1...10 {
+            responseWriter.write(ByteBuffer(string: #"\#(i)\n\r"#))
+            try? await Task.sleep(for: .milliseconds(500))
+        }
+        responseWriter.finish()
+    }
+}
+```
+
+The method is not marked as `throws` as we want the users of this API to handle errors themselves and call `LambdaResponseWriter.reportError(_:)` if needed.
+
+### LambdaRuntime
+
+`LambdaRuntime` is the class that communicates with the Lambda control plane as defined in [Building a custom runtime for AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-custom.html) and forward the invocations to the provided `StreamingLambdaHandler`. It will conform to `ServiceLifecycle.Service` to provide support for `swift-service-lifecycle`.
+
+```swift
+/// The LambdaRuntime object. This object communicates with the Lambda control plane
+/// to fetch work and report errors.
+public final class LambdaRuntime<Handler>: ServiceLifecycle.Service, Sendable
+    where Handler: StreamingLambdaHandler, Handler: ~Copyable
+{
+
+    /// Create a LambdaRuntime by passing a handler, an eventLoop and a logger.
+    /// - Parameter handler: A ``StreamingLambdaHandler`` that will be invoked
+    /// - Parameter eventLoop: An ``EventLoop`` on which the LambdaRuntime will be
+    ///                        executed. Defaults to an EventLoop from
+    ///                        ``NIOSingletons.posixEventLoopGroup``.
+    /// - Parameter logger: A logger
+    public init(
+        handler: consuming sending Handler,
+        eventLoop: EventLoop = Lambda.defaultEventLoop,
```

Comment:
Why are we using our own default EL here instead of the singleton EL?

Reply:
The default EL uses the singleton ELG under the hood. The default EL picks an EL from the singleton ELG. Users can use this to get their invocations onto the same EL as the runtime.

```swift 
try await withTaskExecutor(Lambda.defaultEventLoop) {
    try await runtime.run()
}
```

Reply:
I don't fully buy this yet. If the default becomes the singleton from NIO users can still do that. They just need to call `any()` on the global singleton and then pass around their own EL. How much do we expect people to do this that we need this extra singleton is the question I guess

Reply:
Will we still have the ability to run everything in just one thread (the main thread)? I think for many lambdas you'll never want to fork off a second thread at all. NIO thread & async/await should just run in the main thread (but as global default executor not as `@MainActor`)

Reply:
> I don't fully buy this yet. If the default becomes the singleton from NIO users can still do that. They just need to call `any()` on the global singleton and then pass around their own EL. How much do we expect people to do this that we need this extra singleton is the question I guess

@FranzBusch to do this correctly users would need to do the following:

```swift
let el = NIOSingleton.posix.any()
let runtime = LambdaRuntime(handler: MyHandler(), eventLoop: el)
try await withTaskExecutor(el) {
    try await runtime.run()
}
```

Not sure if explaining this dance is worth it. What is the big problem with having this property. We don't spawn a different ELG. We just make sure that the _same_ EL is used.

> Will we still have the ability to run everything in just one thread (the main thread)? I think for many lambdas you'll never want to fork off a second thread at all. NIO thread & async/await should just run in the main thread (but as global default executor not as `@MainActor`)

Currently I have not planned to keep this functionality. The reason for this is, that users have tripped over this when using Lambda because of course they import some code that uses `DispatchMain` and `@MainActor`. I think the best way out of this would be for NIO to provide an async/await adapted API for this, so that users who are interested in this behavior can easily achieve it:

```swift
try await withCurrentThreadAsEventLoop { el in
    // task executor preference is auto set to el here
    let runtime = LambdaRuntime(handler: MyHandler(), eventLoop: el)
    try await runtime.run()
}
```

@weissi Do you think this would be a reasonable compromise? Make users opt in?

Reply:
> they import some code that uses DispatchMain and @MainActor.

Uuhg, on the server side? That will be super slow, right?

Maybe we could actually make a NIO EL on the main thread host _both_ the MainActor and the global concurrency executor.

> @weissi Do you think this would be a reasonable compromise? Make users opt in?

I do think opt-in is reasonable but we should have tests that indeed everything (async/await code and EL code) runs on the main thread if you opt in.

Reply:
> Maybe we could actually make a NIO EL on the main thread host both the MainActor and the global concurrency executor.

@weissi That would require an Evolution Proposal I assume? So nothing to wait for here, I guess. We will add those tests :)

Reply:
@fabianfett is there nothing like `swift_task_enqueueGlobal_hook` for the main queue on Linux at least?

Reply:
But even without `@MainActor` support, this is still super useful and we should have the tests at least that NIO's I/O code and the user's `async/await` code can run on the same thing.